core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java

```diff
@@ -97,13 +101,23 @@ private static <T> T visitUnion(Type type, Schema union, AvroSchemaWithTypeVisit
         options.add(visit(type, branch, visitor));
       }
     } else { // complex union case
-      int index = 1;
-      for (Schema branch : types) {
-        if (branch.getType() == Schema.Type.NULL) {
-          options.add(visit((Type) null, branch, visitor));
+      Map<Integer, Types.NestedField> icebergFieldMap = new HashMap<>();
+      for (Types.NestedField icebergField : type.asStructType().fields()) {
+        String fieldName = icebergField.name();
+        if (fieldName.equals(UNION_TAG_FIELD_NAME)) {
+          continue;
+        }
+        int fieldId = Integer.valueOf(fieldName.substring(5));
+        icebergFieldMap.put(fieldId, icebergField);
+      }
+
+      for (int i = 0; i < types.size(); ++i) {
+        Schema schema = types.get(i);
+        int fieldIdxInIceberg = types.get(0).getType() == Schema.Type.NULL ? i - 1 : i;
```

Comment:
It seems that `NULL` cannot always be in the first branch? Also something is off if we keep checking on `types.get(0)` in a loop on `types`.